app/utils/IncomeSourcesUtils.scala

```diff
@@ -117,28 +149,57 @@ object IncomeSourcesUtils {
     }
   }
 
-  def getUKPropertyDetailsFromSession(implicit user: MtdItUser[_]): Either[Throwable, CheckUKPropertyViewModel] = {
-    val result: Option[Either[Throwable, CheckUKPropertyViewModel]] = for {
-      tradingStartDate <- user.session.data.get(addUkPropertyStartDate)
-      cashOrAccrualsFlag <- user.session.data.get(addIncomeSourcesAccountingMethod)
-    } yield {
-      Right(CheckUKPropertyViewModel(
-        tradingStartDate = LocalDate.parse(tradingStartDate),
-        cashOrAccrualsFlag = cashOrAccrualsFlag
-      ))
+  def getUKPropertyDetailsFromSession(sessionService: SessionService)(implicit user: MtdItUser[_], ec: ExecutionContext): Future[Either[Throwable, CheckUKPropertyViewModel]] = {
+    sessionService.get(addUkPropertyStartDate).flatMap { startDate: Either[Throwable, Option[String]] =>
+      sessionService.get(addIncomeSourcesAccountingMethod).map { accMethod: Either[Throwable, Option[String]] =>
+        val result = startDate match {
+          case Left(_) => None
+          case Right(dateMaybe) =>
+            accMethod match {
+              case Left(_) => None
+              case Right(methodMaybe) =>
+                for {
+                  foreignPropertyStartDate <- dateMaybe.map(LocalDate.parse)
+                  cashOrAccrualsFlag <- methodMaybe
+                } yield {
+                  CheckUKPropertyViewModel(
+                    tradingStartDate = foreignPropertyStartDate,
+                    cashOrAccrualsFlag = cashOrAccrualsFlag)
+                }
+            }
+        }
+        result match {
+          case Some(propertyDetails) =>
+            Right(propertyDetails)
+          case None =>
+            val errors: Seq[String] = getErrors(startDate, accMethod)
+            Left(new IllegalArgumentException(s"Missing required session data: ${errors.mkString(" ")}"))
+        }
+      }
     }
+  }
 
-    result match {
-      case Some(propertyDetails) =>
-        propertyDetails
-      case None =>
-        val errors: Seq[String] = Seq(
-          user.session.data.get(addUkPropertyStartDate).orElse(Some(MissingKey(s"MissingKey: $addUkPropertyStartDate"))),
-          user.session.data.get(businessTrade).orElse(Some(MissingKey(s"MissingKey: $businessTrade")))
-        ).collect {
-          case Some(MissingKey(msg)) => msg
+  def getErrors(startDate: Either[Throwable, Option[String]], accMethod: Either[Throwable, Option[String]]): Seq[String] = {
+    case class MissingKey(msg: String)
```

Comment:
MissingKey class can be promoted to public and moved into model / errors package as its used in quite a few place as far as I am aware